### PR TITLE
Classifiers

### PR DESCRIPTION
**Pull request by ssanderson**

### CODE REVIEW

File: zipline/pipeline/factors/factor.py [hunk 1]
```diff
@@ -5,20 +5,22 @@
 from operator import attrgetter
 from numbers import Number
 
-from numpy import inf
+from numpy import inf, where, nanmean, nanstd
```
Comment (richafrank) [inline zipline/pipeline/factors/factor.py]: Should we use `nanmean` from our math_utils instead?  It uses bottleneck if it's available.

Comment (ssanderson) [inline zipline/pipeline/factors/factor.py]: yeah, will do

File: zipline/pipeline/classifiers/latest.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    def _validate(self):
+        if self.inputs[0].dtype != int64_dtype:
+            raise TypeError(
+                "{name} expected an input of dtype int64, "
+                "but got {not_bool} instead.".format(
+                    name=type(self).__name__,
+                    not_bool=self.inputs[0].dtype,
+                )
+            )
+        super(Latest, self)._validate()
```
Comment (richafrank) [inline zipline/pipeline/classifiers/latest.py]: I noticed that `Classifier._validate` makes sure to return `super()._validate()`.  Do we need to here too?

File: zipline/lib/normalize.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    >>> naive_grouped_rowwise_apply(data, labels, lambda row: row - row.min())
+    array([[0., 1., 0.],
+           [0., 0., 2.],
+           [0., 0., 0.]])
+    >>> naive_grouped_rowwise_apply(data, labels, lambda row: row / row.sum())
+    array([[ 0.33333333,  0.66666667,  1.        ],
+           [ 0.33333333,  1.        ,  0.66666667],
+           [ 1.        ,  1.        ,  1.        ]])
+    """
+    out = np.empty_like(data)
```
Comment (dmichalowicz) [inline zipline/lib/normalize.py]: Should this be

```python
if out is not None:
    out = np.empty_like(data)
```

File: tests/pipeline/test_factor.py [hunk 4]
```diff
@@ -43,6 +48,20 @@ class F(Factor):
...
+    missing_value = -1
+    inputs = ()
+    window_length = 0
+
+
+class OtherC(Classifier):
+    dtype = int64_dtype
+    missing_value = -1
+    inputs = ()
+    window_length = 0
```
Comment (dmichalowicz) [inline tests/pipeline/test_factor.py]: What's the purpose of having this copy?

File: zipline/pipeline/classifiers/latest.py (hunk 2 above)
Comment (ssanderson) [inline zipline/pipeline/classifiers/latest.py]: Nothing uses these values right now, and all the existing classifiers just return None, but we should probably be consistent.

File: zipline/lib/normalize.py (hunk 3 above)
Comment (ssanderson) [inline zipline/lib/normalize.py]: Yep.

File: tests/pipeline/test_factor.py (hunk 4 above)
Comment (ssanderson) [inline tests/pipeline/test_factor.py]: It's used in `test_normalizations` below for a case where I needed two different classifiers.

**Comment by dmichalowicz** [comment]
@ssanderson Looks good by me. I worked through testing `demean` and `zscore` with different masks and groupby's and they worked well.

**Comment by ssanderson** [comment]
@dmichalowicz thanks!  I'll merge this once I finish making tweaks to the docstrings.

**Pull request merged**